### PR TITLE
feat: maskbook dodo support

### PR DESCRIPTION
**Pull request by AxeAfk**
Maskbook dodo support

### CODE REVIEW

**Comment by septs** [comment]
@AxeAfk 
1. the pr related bounty is ?
2. plz rename the pr title.

File: packages/maskbook/src/plugins/Trader/types/dodo.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    toToken: FungibleTokenDetailed
+    fromAmount: string
+    slippage: number
+    userAddr: string
+    rpc?: string
+    chainId: number
+}
+
+export interface SwapRouteResponse {
+    status: number
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/types/dodo.ts:15]:
```suggestion
    status: 200
```

File: packages/maskbook/src/plugins/Trader/types/dodo.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    useSource: string
+    fromAmount: number
+    value: string
+}
+
+export interface SwapRouteErrorResponse {
+    code?: number
+    status?: number
+    data: string
+}
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/types/dodo.ts:38]:
```suggestion
export type SwapRouteResponse = SwapRouteSuccessResponse | SwapRouteErrorResponse

export interface SwapRouteSuccessResponse {
    code: 200
    status: number
    data: SwapRouteData
}

export interface SwapRouteErrorResponse {
    code: number
    status: number
    data: string
}

export interface SwapRouteData {
    data: string
    priceImpact: number
    resAmount: number
    resCostGas: number
    resPricePerFromToken: number
    resPricePerToToken: number
    targetApproveAddr: string
    targetDecimals: number
    to: string
    useSource: string
    fromAmount: number
    value: string
}
```

File: packages/maskbook/src/resources/DODOIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { SvgIconProps, SvgIcon } from '@material-ui/core'
+
+const svg = (
+    <svg width="20" height="20" viewBox="0 0 40 40" version="1.1">
```
Comment (septs) [inline packages/maskbook/src/resources/DODOIcon.tsx:4]: plz use https://aka.pw/svgomg reduce this svg file size

File: packages/maskbook/src/plugins/Trader/types/dodo.ts (hunk 2 above)
Comment (septs) [reply packages/maskbook/src/plugins/Trader/types/dodo.ts:38]:
```ts
const payload: SwapRouteResponse = await response.json()
if (payload.code !== 200) {
    throw new Error(payload.data ?? 'Unknown Error')
}
/* ... success logic */
```

Comment (septs) [reply packages/maskbook/src/plugins/Trader/types/dodo.ts:38]: see https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions

**Comment by AxeAfk** [comment]
> @AxeAfk
> 
> 1. the pr related bounty is ?
> 2. plz rename the pr title.

1. There is no bounty.
2. Renamed.

File: packages/maskbook/src/plugins/Trader/apis/dodo/index.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    const response = await fetch(`${DODO_BASE_URL}/dodoapi/getdodoroute?${params.toString()}`)
+    const response_: SwapRouteResponse = await response.json()
+
+    const errorResponse = response_ as SwapRouteErrorResponse
+    if (errorResponse.code || errorResponse.status !== 200) throw new Error(errorResponse?.data ?? 'Unknown Error')
+
+    const successResponse = response_ as SwapRouteSuccessResponse
+    return {
+        ...successResponse.data,
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/dodo/index.ts:36]:
```suggestion
    const payload: SwapRouteResponse = await response.json()

    if (payload.status !== 200) {
        throw new Error(payload.data ?? 'Unknown Error')
    }

    return {
        ...payload.data,
```

File: packages/maskbook/src/plugins/Trader/types/dodo.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    chainId: number
+}
+
+export type SwapRouteResponse = SwapRouteSuccessResponse | SwapRouteErrorResponse
+
+export interface SwapRouteSuccessResponse {
+    code: 200
+    status: number
+    data: SwapRouteData
+}
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/types/dodo.ts:20]:
```suggestion
export interface SwapRouteSuccessResponse {
    code: number
    status: 200
    data: SwapRouteData
}
```

**Comment by wenluomask** [comment]
1. Bsc usually works for dodo swap, but RPC error occurred in BNB->Alice swap pair once.

https://user-images.githubusercontent.com/85149605/128134174-6156f5b8-895e-4ef0-b92c-33392b9c091f.mov
Screen Shot 2021-08-04 at 12 19 13 PM
Screen Shot 2021-08-04 at 12 19 04 PM

2. It didn't work for Polygon, but it works in DODO swap website.
Screen Shot 2021-08-04 at 1 36 19 PM
Screen Shot 2021-08-04 at 1 36 06 PM

3. The minimum received currently is as same as estimated received. They should be different like DODO swap website.
Screen Shot 2021-08-04 at 2 26 17 PM
Screen Shot 2021-08-04 at 2 26 10 PM

4. If possible, add Liquidity Provider Fee and View pair analytics session.
Screen Shot 2021-08-04 at 2 56 05 PM

**Comment by AxeAfk** [comment]
@wenluomask  Thanks for helping with the test.

About 1 and 2, maybe it was because I didn't use Hook to get the data yesterday. This was fixed in the last commit, and I tested 1 and 2 locally and there is no problem. Please try again with latest commit.

About 3, fixed the calculation formula in latest commit.

About 4, DODO may have multiple fund pool jumps on the transaction path, so there is no details about the Liquidity Provider Fee. The estimated amount is the amount after deducting the handling fee of each fund pool.

**Comment by wenluomask** [comment]
Thanks for your reply. I tested the latest commit, but bug #2 is still exist. You can check the following video:
https://user-images.githubusercontent.com/85149605/128291385-601426d5-4456-4422-92bc-29e559644d1b.mov

For bug #4, if possible, please add "view pair analytics" feature for DODO. For example, if we are doing Matic->Alice swap pair, "view pair analytics" will link to https://app.dodoex.io/exchange/BNB-ALICE?network=bsc-mainnet .

**Comment by wenluomask** [comment]
In addition, the minimum received in mask is different from DODO swap website. Is it possible to be the same?

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 6]
```diff
@@ -97,6 +97,22 @@ export function resolveTradePairLink(tradeProvider: TradeProvider, address: stri
             return `https://v2.info.uniswap.org/pair/${address}`
         case TradeProvider.ZRX:
             return ''
+        case TradeProvider.DODO:
+            switch (networkType) {
+                case NetworkType.Ethereum:
+                    return `https://app.dodoex.io/exchange/${address}?network=mainnet&forced=true`
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:103]: plz use https://npm.im/urlcat build your urls.

```ts
urlcat('https://app.dodoex.io/exchange/:address', {
    address,
    network: networkNames[networkType],
    forced: true
})

const networkNames: Record<NetworkType, string> = {
    [NetworkType.Ethereum]: 'mainnet',
    // more remapping ...
}
```

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 7]
```diff
@@ -97,6 +97,22 @@ export function resolveTradePairLink(tradeProvider: TradeProvider, address: stri
...
+                    return `https://app.dodoex.io/exchange/${address}?network=bsc-mainnet&forced=true`
+                case NetworkType.Polygon:
+                    return `https://app.dodoex.io/exchange/${address}?network=matic&forced=true`
+                case NetworkType.Arbitrum:
+                    return ''
+                default:
+                    safeUnreachable(networkType)
+                    return ''
+            }
+        case TradeProvider.DODO:
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:114]: why duplicate switch case?

**Comment by AxeAfk** [comment]
The error under the Matic network has been resolved, and the view pair analytics link has been added.

The minimum received calculation results I see here are the same. Is it because of the different slippage setting parameters that the results are different?

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 8]
```diff
@@ -98,21 +100,11 @@ export function resolveTradePairLink(tradeProvider: TradeProvider, address: stri
...
-                    return ''
-                default:
-                    safeUnreachable(networkType)
-                    return ''
-            }
-        case TradeProvider.DODO:
-            return `https://app.dodoex.io/exchange/${address}`
+            return urlcat('https://app.dodoex.io/exchange/:address', {
+                address,
+                network: networkNames[networkType],
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:105]: you need check `networkNames` collection include `networkType`

e.g:
```
if (!(networkType in networkNames)) {
    console.error('... error message ...')
    return ''
}
```

Comment (septs) [reply packages/maskbook/src/plugins/Trader/pipes.ts:106]: plz output error message, thanks

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 9]
```diff
@@ -115,6 +132,8 @@ export function resolveTradePairLink(tradeProvider: TradeProvider, address: stri
             return `https://info.quickswap.exchange/pair/${address}`
         case TradeProvider.PANCAKESWAP:
             return `https://pancakeswap.info/pool/${address}`
+        case TradeProvider.DODO:
+            return `https://app.dodoex.io/pool/list/${address}`
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:136]: plz use urlcat.

File: packages/maskbook/src/plugins/Trader/apis/dodo/index.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    if (request.fromToken) {
+        params.set('fromTokenAddress', request.fromToken.address)
+        params.set('fromTokenDecimals', String(request.fromToken.decimals))
+    }
+    if (request.toToken) {
+        params.set('toTokenAddress', request.toToken.address)
+        params.set('toTokenDecimals', String(request.toToken.decimals))
+    }
+
+    const response = await fetch(`${DODO_BASE_URL}/dodoapi/getdodoroute?${params.toString()}`)
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/dodo/index.ts:28]: plz use urlcat build the urls

File: packages/maskbook/src/plugins/Trader/apis/dodo/index.ts [hunk 11]
```diff
@@ -8,24 +8,22 @@ import type {
...
-        params.set('fromTokenDecimals', String(request.fromToken.decimals))
-    }
-    if (request.toToken) {
-        params.set('toTokenAddress', request.toToken.address)
-        params.set('toTokenDecimals', String(request.toToken.decimals))
-    }
-
-    const response = await fetch(`${DODO_BASE_URL}/dodoapi/getdodoroute?${params.toString()}`)
+    const response = await fetch(
+        urlcat(`${DODO_BASE_URL}/dodoapi/getdodoroute`, {
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/dodo/index.ts:15]:
```suggestion
        urlcat(DODO_BASE_URL, '/dodoapi/getdodoroute', {
```

**Comment by wenluomask** [comment]
I tried to swap Matic to 0xBTC, but RPC error occurs, here is the video I recorded:

https://user-images.githubusercontent.com/85149605/128384131-c20706e2-9ad7-429c-9705-e53cca39c956.mov

**Comment by guanbinrui** [comment]
Replaced by #3882.

**Pull request closed without merging**